### ISSUE TITLE
In-place operations are reordered across the forward-backward in autograd function

### ISSUE THREAD

**Issue by zxfseserer** [pytorch/pytorch#154820]
### 🐛 Describe the bug

When performing in-place operations in an autograd function, the forward/backward update end up getting reordered so that the backward update is done simultaneously during the forward, or if you implicate a gradient, flip it so that the forward in-place is done only during the backward.

Putting the update in a torch.lib custom op had no effect.
```
import torch


class Func(torch.autograd.Function):
    @staticmethod
    def forward(ctx, dummy, inplace_tensor, attach_gradient):
        inplace_tensor.add_(1)

        ctx.attach_gradient = attach_gradient
        ctx.inplace_tensor = inplace_tensor
        return dummy.clone()

    @staticmethod
    def backward(ctx, grad_output):
        inplace_tensor = ctx.inplace_tensor
        attach_gradient = ctx.attach_gradient

        if attach_gradient:
            gradient_attachment = (grad_output * 0 + 1)
            inplace_tensor.add_(1 * gradient_attachment)
        else:
            inplace_tensor.add_(1)


        return grad_output, None, None


def call(dummy, inplace_tensor, attach_gradient):
    return Func.apply(dummy, inplace_tensor, attach_gradient)

compiled_call = torch.compile(call)
dummy = torch.randn((2,), requires_grad=True).to('cuda')
inplace_tensor = torch.zeros((2,), requires_grad=False).to('cuda')

print(f'Compiled no gradient attachment')
loss = compiled_call(dummy, inplace_tensor, False).sum()
print(f'Pre backward inplace: {inplace_tensor}')
loss.backward()
print(f'Post backward inplace: {inplace_tensor}\n')
inplace_tensor.zero_()

print(f'Compiled with gradient attachment')
loss = compiled_call(dummy, inplace_tensor, True).sum()
print(f'Pre backward inplace: {inplace_tensor}')
loss.backward()
print(f'Post backward inplace: {inplace_tensor}\n')
inplace_tensor.zero_()

print(f'Uncompiled')
loss = call(dummy, inplace_tensor, False).sum()
print(f'Pre backward inplace: {inplace_tensor}')
loss.backward()
print(f'Post backward inplace: {inplace_tensor}\n')
inplace_tensor.zero_()
```

produces:

Compiled no gradient attachment
Pre backward inplace: tensor([2., 2.], device='cuda:0')
Post backward inplace: tensor([2., 2.], device='cuda:0')

Compiled with gradient attachment
Pre backward inplace: tensor([0., 0.], device='cuda:0')
Post backward inplace: tensor([2., 2.], device='cuda:0')

Uncompiled
Pre backward inplace: tensor([1., 1.], device='cuda:0')
Post backward inplace: tensor([2., 2.], device='cuda:0')

@yf225 @zou3519 

### Versions

Versions of relevant libraries:
[pip3] numpy==2.1.2
[pip3] nvidia-cublas-cu11==11.11.3.6
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu11==11.8.87
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu11==11.8.89
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu11==11.8.89
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu11==9.1.0.70
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu11==10.3.0.86
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu11==11.4.1.48
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu11==11.7.5.86
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu11==2.21.5
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu11==11.8.86
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] pytorch-triton==3.3.0+git96316ce5
[pip3] torch==2.7.0
[pip3] torchaudio==2.6.0.dev20250513+cu118
[pip3] torchvision==0.22.0.dev20250513+cu118
[pip3] triton==3.3.0

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @amjames @bdhirsh

**Comment by anijain2305** [comment]
@zou3519 Wondering I might need your help in debugging this one to understand the expected torch.compile behavior. 

Turning to high priority for now because of possible accuracy mismatch.

**Comment by zou3519** [comment]
@bdhirsh I remember there was some PR that allowed more input mutation in autograd.Function, is this related?

**Comment by bdhirsh** [comment]
@IvanKobzarev this looks something that should have been fixed by https://github.com/pytorch/pytorch/pull/129130, can you take a look?

**Comment by IvanKobzarev** [comment]
There are 2 separate problems.
1/ we "glue" mutations of the same primal that happen in forward and backward in one copy_.
2/ copy_(src, constant_node_prop) with src as constant_propagation is not captured by partitioner in forward or backward - extracted to a separate issue https://github.com/pytorch/pytorch/issues/155242

**Comment by IvanKobzarev** [comment]
@zxfseserer If you are blocked by this issue - 2 PRs are out to resolve it:

https://github.com/pytorch/pytorch/pull/155353 - landed to main recently
https://github.com/pytorch/pytorch/pull/155354 - On review , not landed yet

**Comment by IvanKobzarev** [comment]
Fixed by https://github.com/pytorch/pytorch/pull/155353  and  https://github.com/pytorch/pytorch/pull/155354 that was landed recently
